go-controller/pkg/level_driven_controller/level_driven_controller.go

```diff
@@ -0,0 +1,271 @@
+			}
+		}()
+	}
+
+	// add shutdown goroutine waiting for c.stopCh
+	go func() {
+		// wait until we're told to stop
+		<-stopCh
+		c.cleanup()
+	}()
+
+	return nil
+}
+
+func (c *controller[T]) cleanup() {
+	c.queue.ShutDown()
+	if c.eventHandler != nil {
+		if err := c.informer.RemoveEventHandler(c.eventHandler); err != nil {
+			klog.Errorf("Failed to remove event handler for controller %s: %w", c.name, err)
+		}
+	}
+}
+
+func (c *controller[T]) onAdd(objInterface interface{}) {
+	newObj, ok := objInterface.(*T)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("controller %s: expecting %T but received %T", c.name, *new(T), newObj))
+		return
+	}
+	if !c.handler.ObjNeedsUpdate(nil, newObj) {
+		return
+	}
+
+	key, err := cache.MetaNamespaceKeyFunc(newObj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("controller %s: couldn't get key for object %+v: %v", c.name, newObj, err))
+		return
+	}
+	c.queue.Add(key)
+}
+
+func (c *controller[T]) onUpdate(oldObjInterface, newObjInterface interface{}) {
+	oldObj, ok := oldObjInterface.(*T)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("controller %s: expecting %T but received %T", c.name, *new(T), oldObj))
+		return
+	}
+	newObj, ok := newObjInterface.(*T)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("controller %s: expecting %T but received %T", c.name, *new(T), newObj))
+		return
+	}
+
+	if !c.handler.ObjNeedsUpdate(oldObj, newObj) {
+		return
+	}
+
+	key, err := cache.MetaNamespaceKeyFunc(newObj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("controller %s: couldn't get key for object %+v: %v", c.name, newObj, err))
+		return
+	}
+	c.queue.Add(key)
+}
+
+func (c *controller[T]) onDelete(objInterface interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(objInterface)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("controller %s: couldn't get key for object %+v: %v", c.name, objInterface, err))
+		return
+	}
+	c.queue.Add(key)
+}
+
+// processNextQueueItem returns false when the queue is shutdown and the handling should be stopped.
+// Otherwise, it handles the next item from the queue and always returns true.
+func (c *controller[T]) processNextQueueItem() bool {
+	key, shutdown := c.queue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	defer c.queue.Done(key)
+
+	err := c.handler.Reconcile(key.(string))
+	if err != nil {
+		if c.queue.NumRequeues(key) < maxRetries {
+			klog.Infof("Controller %s: error found while processing %s: %v", c.name, key.(string), err)
+			c.queue.AddRateLimited(key)
+			return true
+		}
+		klog.Warningf("Controller %s: dropping %s out of the queue: %v", c.name, key.(string), err)
+		utilruntime.HandleError(err)
+	}
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *controller[T]) AddToQueue(key string) {
+	c.queue.Add(key)
+}
+
+// StoppableController is a level-driven controller that is shut down after Stop() call.
+// It creates its own stop channel and waiting group to ensure proper cleanup.
+type StoppableController[T any] struct {
```

Comment:
in ovn-k we have 2 ways to stop a controller: with stop channel or with Stop() call. These options require different Start() signatures, so to allow multiple `Start` options, I made these 2 types (there could be a third one using context in the future). To reuse start functionality, I had to create a common `start` method, and then 2 wrapper types to call it with different args.

Reply:
I think we should have a single way for this API and users should adapt to it. So instead of encouraging all the different ways you can handle this, let's just go for one.

Reply:
let me try to clarify what i meant with this package
It is not supposed to be a better way to write controllers, but just a way to not copy the same code multiple times, mostly to avoid bugs. That is why in this package I am trying to adapt to how controllers are used right now (that includes implementing Stop() methods making sure the channel is closed and Wait() call is there) instead of defining the nice interface and then catching bugs from the user side.
I find it useful even if it is not perfect, since it decreases code duplication, and helps you use the controller in a proper way (e.g. you can find some strange combinations of wait group, stop channel and Stop method usages, so this interface is supposed to help with that)

Reply:
The only controllers using this utility are the ones introduced in this PR and they just use `NewStoppableController`.
 
Since it is only that, you wouldn't need a new type (unless you want an interface). I also think we should start to avoid if possible stop channels and moving wait groups around. So I would just have a `Run(ctx Context, ...)` that blocks. Why would we encourage different ways of doing it? We should consolidate upstream users to do it in the same way, and If it can't be avoided, they can have their own adapters. And if it can't be avoided, those adapters can also be utilities in their own package with their own doc. 

`NewStopChanController` is just used in the test so why not move it to the test?

It is very confusing to export `NewStopChanController` and `NewStoppableController`

I am just trying to help, within reason, but if you are not convinced then I guess I failed and we can just move on with what you have.